Chapter: Evaluation / Section: Results (#29)
ysndr/MSc-Thesis: chapter/evaluation.md

### Quantitative

## Results

### Process
### Qualitative

As outlined in [#sec:qualitative-study-outline], the qualitative study consists of two parts conducted before and after an introductory workshop.
The pre-evaluation aimed to catch the users's expected features and behaviours, while the main survey asked users about their concrete experiences with the NLS.

#### Pre-Evaluation

Responding to the first point (c.f. [#sec:expected-features]), the participants unanimously identified four of the six foundational language server capabilities that guided the implementation of the project (c.f. [@sec:commands-and-notifications, @langserverorg] );

Type-information on hover
  ~ was named almost uniformly.
    The participants showed a special interest in this feature describing specific behaviours.

>     The desired information exposed by this feature are value types including applied contracts and documentation as well as function signatures.

Reviewer: ```suggestion
    The desired information exposed by this feature are types (including applied contracts) and documentation as well as function signatures.
```

Are types and function signatures really different?

Author: I dont think so but is was mentioned explicitly.
As in: for functions - function signatures and documentation

  ~ When asked about the hover LSP method in particular, participants name additional function documentation, default values and the visualization of scopes as an additional features.

Diagnostics
  ~ are widely understood as an important feature.
    Participant had very particular opinions about the behavior and detail of diagnostics including error message at the correct location in the code signaling syntax errors or possibly evaluation errors and contract breaches.
    In either case the diagnostic should be produced "On-the-fly" while typing or upon saving the document.
  ~ When asked about the diagnostics feature of language servers directly, the answers corroborated these initial opinions.
    In addition some participants named code linting, i.e. warnings about code style, unused variables, deprecated code and undocumented elements, as well as structural analysis hints as possible features.

>     Structural analysis was imagined to go that far as being able to "suggest how to fix" mistakes in the code.

Reviewer: ```suggestion
    Structural analysis was imagined to go as far as being able to "suggest how to fix" mistakes in the code.
```

Code Completion

>   ~ was equally name in all but one response.

Reviewer: ```suggestion
  ~ was equally named in all but one response.
```

    It was described as a way to chose from possible completion candidates 
    of options.
    The answers included aspirational vague descriptions of such a feature including the a way to automatically prioritize specific items.
  ~ Responding about the concrete LSP feature, participants listed variables, record fields, types, functions and function argument candiates as possible completion candidates.
    Moreover, some suggested the inclusion of the completion context to guide priorization as well as auto-generated contract and function skeletons.

Jump-to-Definition
   ~ was included in three fourth of responses.

>    ~ The specific feature survey revealed the exoected behaviour in more detail;

Reviewer: ```suggestion
   ~
```

Not a very useful sentence

Neither of them shows clear trends with positive and negative reults distributed almost evenly between positive and negative sentiments.

The first graph ([@fig:results-comp-experience]) represents the participants' general experience with the relevant features.
It shows that each feature worked without issue in at least one instance.
Yet, three features were reported to not work at all and no feature  left the users unsurprised.
Participant found the hover and diagnostic features to behave particularly unexpectedly. 

For the second item of each feature, the survey asked the subjects to rate the quality of the language server based on their expectations.
[Figure @fig:results-comp-expectations] summarizes the results.
Apart from the same three occasions in which a feature did not work for one participant, the majority of responses show that NLS met its user's expectations at least partially.
The results are however highly polarized as the Jump-to-Definition and Hover features demonstrate; Each received equally many votes for being inapt and fully able to hold up to the participants expectations at the same time.

> Other features were left with with a nuniformly distributed assessment (e.g. Completion and Find-References).

Reviewer: ```suggestion
Other features were left with with an uniformly distributed assessment (e.g. Completion and Find-References).
```

The clearest result was achieved by the Diagnostics feature, which received a slight but uncontended positive sentiment.

> Asking about the general satisfaction with each feature, results in the same mixed answers as seen in [@fig:results-comp-satisfaction].

Reviewer: ```suggestion
Asking about the general satisfaction with each feature results in the same mixed answers as seen in [@fig:results-comp-satisfaction].
```

More practically, cross file navigation was named an important missing feature.

##### General Performance {#sec:general-performance@res}

The responses to this item suggest that NLS' perfomance is largely dependent on its usage.
On unmodified files queries were reported to evaluate "instantaniously".
However modifying files caused that "modifications stack up" causing high CPU usage and generally "very slow" responses.
Others pointed out that documentation was slow to resolve while the server itself was "generally fast".

### Quantitative

The quantitative evaluation focuses on the performance characteristics of NLS.
As described in [@sec:eval-methods-quantitative] a tracing module was embedded into the NLS binary which recorded the runtime together with the size of the analyzed data, i.e., the number of linearization items [@sec:linearization] or size of the analyzed file.

#### Dataset

The underlying data set consists of 16760 unique trace records.
Since the `textDocument/didOpen` method is executed on every update of the source, it greatly outnumbers the other events.
The final distrubution of methods traced is:

> | Method                    | count | linearization based |
> | ------------------------- | ----- | ------------------- |
> | `textDocument/didOpen`    | 13436 | no                  |
> | `textDocument/completion` | 2981  | yes                 |
> | `textDocument/hover`      | 227   | yes                 |
> | `textDocument/definition` | 68    | yes                 |
> | `textDocument/references` | 49    | yes                 |

Author: ```suggestion
| Method                    | count | linearization based |
| ------------------------- | ----- | ------------------- |
| `textDocument/didOpen`    | 13436 | no                  |
| `textDocument/completion` | 2981  | yes                 |
| `textDocument/hover`      | 227   | yes                 |
| `textDocument/definition` | 68    | yes                 |
| `textDocument/references` | 49    | yes                 |
| ------------------------- | ----- | ------------------- |
| total                     | 16761 |                     |
```

: Number of traces per LSP method

![Distribution of linearization based LSP requests](log_analysis/figures/distribution-by-lin-size.svg){#fig:distribution-by-lin-size}

![Distribution of file analysis requests](log_analysis/figures/distribution-by-file-size.svg){#fig:distribution-by-file-size}

[Figures @fig:distribution-by-lin-size, @fig:distribution-by-file-size] break up these numbers by method and linearization size or file size respectively.
The first figure shows a peak number of traces for completion events between $0$ to $1$ linearization items as well as local maxima around a linearization size of $20$ to $30$ and sustained usage of completion requests in files of $90-400$ items.

> Similar to the completion requests (but well outnumbered in total counts), other other methods were used mainly in the range between $200$ and $400$ linearization items.

Reviewer: ```suggestion
Similar to the completion requests (but well outnumbered in total counts), other methods were used mainly in the range between $200$ and $400$ linearization items.
```

Reviewer: Maybe a reminder, even quick, about what exactly means "linearization size/linearization items" could be useful here. I don't know where this chapter will be in the final manuscript, but the reader may not necessarily have still a precise idea of what all of those are.

Finally, document linearization as associated with the `textDocument/didOpen` method shows a great range with maxima of $1.5 * 10^5µs$ (about $2.5$ minutes) and a generally greater inter quartile range spanning more than tow orders of magnitude.

#### Special cases

<div id="fig:correlation-linearization-methods">
![Runtime latencies of completion requests at differnet linearization sizes](log_analysis/figures/correlation-completions.svg){#fig:correlation-completions width=50%}
![Runtime latencies of hover requests at differnet linearization sizes](log_analysis/figures/correlation-hovers.svg){#fig:correlation-hovers width=50%}

![Runtime latencies of find-references requests at differnet linearization sizes](log_analysis/figures/correlation-references.svg){#fig:correlation-references width=50%}
![Runtime latencies of jump-to-definition requests at differnet linearization sizes](log_analysis/figures/correlation-definitions.svg){#fig:correlation-definition width=50%}

Runtime latencies of different linearization based methods 
</div>

![Runtime latencies of file update handlings at different file sizes](log_analysis/figures/correlation-opens.svg){#fig:correlation-opens width=50%}

> Setting the runtime of completion requests in relation to the linearization size on which the command was performed, shows no clear correlation between the dimensions.

Reviewer: ```suggestion
Setting the runtime of completion requests in relation to the linearization size on which the command was performed shows no clear correlation between the dimensions.
```

In fact the correlation coefficient between both variables measures $0.01617$ on a linear scale and $0.26$ on a $\log_{10}\log_{10}$ scale.
Instead, vertical colums stand out in the correlation graph [@fig:correlation-completions].
The height of these colums varies from one to five orders of magnitude.

> Considering the item density shows that especially high columns form whenever the server receives a higher load of requests.

Reviewer: ```suggestion
The item density shows that especially high columns form whenever the server receives a higher load of requests.
```

#### Dataset

The underlying data set consists of 16760 unique trace records.
Since the `textDocument/didOpen` method is executed on every update of the source, it greatly outnumbers the other events.
The final distrubution of methods traced is:

| Method                    | count | linearization based |
| ------------------------- | ----- | ------------------- |
| `textDocument/didOpen`    | 13436 | no                  |
| `textDocument/completion` | 2981  | yes                 |
| `textDocument/hover`      | 227   | yes                 |
| `textDocument/definition` | 68    | yes                 |
| `textDocument/references` | 49    | yes                 |
| ------------------------- | ----- | ------------------- |
| total                     | 16761 |                     |

: Number of traces per LSP method

![Distribution of linearization based LSP requests](log_analysis/figures/distribution-by-lin-size.svg){#fig:distribution-by-lin-size}

![Distribution of file analysis requests](log_analysis/figures/distribution-by-file-size.svg){#fig:distribution-by-file-size}

> [Figures @fig:distribution-by-lin-size, @fig:distribution-by-file-size] break up these numbers by method and linearization size or file size respectively.

Author: ```suggestion
[Figures @fig:distribution-by-lin-size, @fig:distribution-by-file-size] break up these numbers by method and linearization size or file size respectively.
The linearization is the linear representation of an enriched AST.
It is explained in great detail in [@sec:linearization]. 
```
